drivers: clock_control: nrf: Fix error during initialization

nordic-krch opened the pull request:
In certain configurations, calibration may be started before kernel is
up and running. That lead to error because calibration is using kernel
api (work queue). Issue was observed if low frequency clock was started
before kernel initialization.

Moved calibration initialization to application phase because system
work queue is initialized in post kernel phase.

Fixes #25568.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

anangl commented on drivers/clock_control/nrf_power_clock.c:
```diff
@@ -424,11 +420,11 @@ void nrf_power_clock_isr(void *arg)
 
 	if (clock_event_check_and_clean(NRF_CLOCK_EVENT_LFCLKSTARTED,
 					NRF_CLOCK_INT_LF_STARTED_MASK)) {
+		clkstarted_handle(dev, CLOCK_CONTROL_NRF_TYPE_LFCLK);
 		if (IS_ENABLED(
 			CONFIG_CLOCK_CONTROL_NRF_K32SRC_RC_CALIBRATION)) {
 			z_nrf_clock_calibration_lfclk_started();
 		}
-		clkstarted_handle(dev, CLOCK_CONTROL_NRF_TYPE_LFCLK);
```
I don't entirely understand the motivation of this change. I thought the calibration was to be started ASAP after the LF clock gets running, before any user callbacks. Why was the order inverted?

nordic-krch replied:
the problem is that calibration is using system work queue which is initialized in post kernel so when LF clock is started before system work queue is ready then error occurs. Configuration provided in the issue was provoking this case.

I don't like this solution and it would be better to not rely on work queue but for that #21667 would need to be merged and that won't happen for that release.

anangl replied:
Yes, I understand the issue and the idea behind this PR.
My comment was only about moving `clkstarted_handle(dev, CLOCK_CONTROL_NRF_TYPE_LFCLK)` before `z_nrf_clock_calibration_lfclk_started()`. Is that actually needed?

nordic-krch replied:
handling of lfclk started event changes state of the clock. If `z_nrf_clock_calibration_lfclk_started` is called earlier then it won't start calibration because clock is not yet started. If init was executed earlier then calibration will never be started.

anangl replied:
But `z_nrf_clock_calibration_lfclk_started` does not check the state of the clock. It only needs the `clock_dev` pointer to start the calibration. And this pointer is set in `z_nrf_clock_calibration_init` regardless of whether the clock is running or not. So I still don't see the need of this inversion of the order.

pabigot reviewed (commented):
If the only thing we need the work queue for during calibration is the temperature, why not just delay initializing `temp_sensor` until we reach the `APPLICATION` init level?  From what I can tell that will prevent entering a code path that requires the kernel.

Then we can have the clock calibration start early, if not as early as proposed in #22758.

carlescufi commented:
@nordic-krch and @anangl ping on this one.

nordic-krch commented:
@pabigot 

> If the only thing we need the work queue for during calibration is the temperature, why not just delay initializing temp_sensor until we reach the APPLICATION init level? From what I can tell that will prevent entering a code path that requires the kernel.

That will not delay usage of `k_work_queue` since it is only skipped if `temp_sensor` handle is null and fetching device handle before it is being initialized will return valid handle.

pabigot commented:
> That will not delay usage of `k_work_queue` since it is only skipped if `temp_sensor` handle is null and fetching device handle before it is being initialized will return valid handle.

That's exactly my point.  Move:
```
        if (CONFIG_CLOCK_CONTROL_NRF_CALIBRATION_MAX_SKIP != 0) {
                temp_sensor = temp_device();
        }
```
from `z_nrf_clock_calibration_init()` to a `SYS_INIT` function that runs after this one from `kernel/system_work_q.c`:

    SYS_INIT(k_sys_work_q_init, POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);

nordic-krch committed:
drivers: clock_control: nrf: Fix error during initialization

Prevent usage of system work queue before it is ready by delaying
initialization of temperature device. Work queue is not used until
application stage.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@pabigot applied your suggestion.

carlescufi reviewed (approved):
Very nice solution.

carlescufi commented:
Note: I also verified that this runs correctly with the sample that was provided in #25568.
